### ISSUE TITLE
Automatically retry failed deployments

### ISSUE THREAD

**Issue by jwendell** [openshift/origin#7522]
I'm in a situation where the readiness probe is failing forever.

AFAIK it's only restarted if the liveness probe fails, not readiness, right?

So, this pod never starts. Should openshift have also an option to restart the pod when readiness probe fails for, say, 5 tries?

**Comment by mfojtik** [comment]
@jwendell i do believe openshift/kube restart the pod when the readiness pod fails if the pod is part of the replication controller (in that case when readiness fails, RC will automatically start new pod).

**Comment by jwendell** [comment]
That didn't happen to me. It waits infinitely and didn't start a new pod...

**Comment by mfojtik** [comment]
@jwendell i was wrong, sorry for confusion, when pod is not ready, it gets removed from the service but it wont  restart.

**Comment by mfojtik** [comment]
@jwendell what you can do is:

1) Define readiness probe and set `timeoutSeconds`
2) Define liveness probe and set `initialDelaySeconds` to equal timeout seconds from readiness probe

What I think will happen is that when readiness probe fails in given timeoutSeconds the liveness probe runs and restarts the container in Pod. 

@smarterclayton does that sounds right?

**Comment by jwendell** [comment]
I think (at least the doc says) that it doesn't work like you described.

When readiness probe fails (due to a timeout or because the script exits with a non-zero status) openshift/kube doesn't recognize it as ready and probes again, infinitely.

The liveness probe is never executed because the pod never reached the "alive" status.

**Comment by smarterclayton** [comment]
Hrm, liveness probes and readiness probes should be orthogonal.  The
*deployment* should time out eventually and roll back, but liveness should
restart the pods at any point.

**Comment by pweil-** [comment]
If this is in the context of a deployment then you will need to set the rolling parameters to specify your tolerance for downtime.  For instance if you want no downtime you can use something like

```
"spec": {
        "strategy": {
          "resources": {},
          "rollingParams": {
             "maxUnavailable": 0
          }
        },
```
which would result in a deployer taking action like this for a dc that contained a failing readiness probe:

```
oc logs -f hello-openshift-2-deploy
I0224 21:50:56.546280       1 deployer.go:201] Deploying from default/hello-openshift-1 to default/hello-openshift-2 (replicas: 1)
I0224 21:50:57.576967       1 rolling.go:228] RollingUpdater: Continuing update with existing controller hello-openshift-2.
I0224 21:50:57.592071       1 rolling.go:228] RollingUpdater: Scaling up hello-openshift-2 from 0 to 1, scaling down hello-openshift-1 from 1 to 0 (keep 1 pods available, don't exceed 2 pods)
I0224 21:50:57.592107       1 rolling.go:228] RollingUpdater: Scaling hello-openshift-2 up to 1
F0224 22:01:00.993431       1 deployer.go:69] timed out waiting for any update progress to be made
```

Note, the timeout is 10 minutes by default.  

/cc @kargakis 

If this is not within the context of a deployment then the liveness probe is what would trigger the restart.

**Comment by jwendell** [comment]
Thanks @pweil- for the hint about "rolling" strategy. I was using "recreate".

Using rolling I can pass the timeout parameter. Openshift respects that and, if the pod is not ready within that timeout, the deployment fails.

What I need now is a way to automatically retry a failed deployment. Something like: Try to deploy this thing for 5 times, then, give up. Is there such feature in Openshift?

**Comment by kargakis** [comment]
@jwendell you can run `oc deploy DCNAME --retry` and your deployment will be retried.

**Comment by jwendell** [comment]
@kargakis yep, but I'd like it to be automatically. Once a deploy fails, retry automatically. (I don't have manual control over our environment). Is that possible?

**Comment by kargakis** [comment]
@jwendell no, there is no such thing as an automatic retry right now. A failed deployment most of the time is due to a permanent error that requires action.

**Comment by smarterclayton** [comment]
If you want to try even longer, you can use longer deployment timeouts.

On Thu, Feb 25, 2016 at 10:36 AM, Michail Kargakis <notifications@github.com

**Comment by kargakis** [comment]
@smarterclayton ProgressDeadlineSeconds will be configurable for upstream Deployments, meaning that nil will always retry to deploy. We could do the same for deployment configs but with the current default being the default?

**Comment by smarterclayton** [comment]
Sure

**Comment by kargakis** [comment]
ProgressDeadlineSeconds API will be merged soon in Kube and we should have the implementation merged by Kube 1.5. Adding this to the 1.5 milestone for Origin.

**Comment by kargakis** [comment]
Technically we cannot guarantee that the deployer pod will be up forever and we cannot replace a dead deployer with a new one w/o keeping it simple. It's easier to just use upstream Deployments in this case. In the future, we will provide upstream Deployments with all of the features of DeploymentConfigs, hopefully layered on top in a composable way.
